BUG: Fix all kinds of problems when itemsize == 0

eric-wieser opened the pull request:
Split off all of the bugfixes from #8970.

eric-wieser committed:
BUG: Prevent hang when sorting dtype=S0 arrays

eric-wieser committed:
BUG: Prevent ValueError when resizing V0 arrays

eric-wieser committed:
BUG: Allow empty strings to be viewed as themselves

eric-wieser committed:
BUG: Allow empty strings to be pickled

eric-wieser committed:
MAINT: Pre-emptively handle the case when elsize == 0

Fixes np.fromiter and np.dtype. Neither of these code paths can currently be hit.

juliantaylor commented on numpy/core/src/multiarray/ctors.c:
```diff
@@ -3711,14 +3711,15 @@ PyArray_FromIter(PyObject *obj, PyArray_Descr *dtype, npy_intp count)
-            if (elcount <= NPY_MAX_INTP/elsize) {
-                new_data = PyDataMem_RENEW(PyArray_DATA(ret), elcount * elsize);
+            if (!npy_mul_with_overflow_intp(&nbytes, elcount, elsize)) {
+                new_data = PyDataMem_RENEW(PyArray_DATA(ret), nbytes);
```
the else clause and the `new_data` variable can be removed

eric-wieser replied:
What do you mean by removing the variable?

Fixing that feels kinda out of scope for this patch

charris replied:
IIRC, realloc will free the memory if size is zero, which we don't want to do. Worth checking, normally we keep a minimum of 1 byte to avoid problems.

eric-wieser replied:
Sometimes, yes - but we don't actually care. All that matters is if `realloc` returns NULL, (which `realloc(..., 0)` does on some platforms, in which case we fail with a memory error.

Either way, we don't need to worry about this yet - this function has a check for elsize == 0, which we shouldn't remove unless #8970 is merged.

So yes, technically, this file didn't need a patch at all - but I'd argue that `npy_mul_with_overflow_intp` does a better job of communicating the intention

juliantaylor commented on numpy/core/src/multiarray/shape.c:
```diff
@@ -126,22 +121,21 @@ PyArray_Resize(PyArrayObject *self, PyArray_Dims *newshape, int refcheck,
-            optr = PyArray_BYTES(self) + oldsize*elsize;
-            n = newsize - oldsize;
+            optr = PyArray_BYTES(self) + oldnbytes;
+            n = (newnbytes - oldnbytes) / elsize;
```
this looks scetchy as the == 0 check is removed.
It might not go in here as then newbytes and oldnbytes are 0 but it's still dangerous

eric-wieser replied:
The idea here is that if the number of bytes are different, the element must be non empty. I guess rather than doing the division, we could just calculate this directly though

juliantaylor replied:
hm its probably good as is.

eric-wieser replied:
Patch incoming, I think it makes it cleaner

eric-wieser replied:
Also uncovered another unrelated bug

juliantaylor commented on numpy/core/src/multiarray/getset.c:
```diff
@@ -472,25 +472,19 @@ array_descr_set(PyArrayObject *self, PyObject *arg)
+     * don't allow np.void to be duplicated
+     */
+    if (newtype->type_num == NPY_VOID && newtype->elsize == 0
+                                      && PyArray_DESCR(self)->elsize != 0) {
```
our usual style is having the operator on the line before but indenting the next line

eric-wieser replied:
That doesn't match with what I remember seeing, eg in `mapping.c`. Operators should be at the start of the line, so that they're not missed when reading through.

I can wrap after `NPY_VOID` if you'd like, but I'm not sure what level to indent to as a result

charris replied:
The usual style (Python recommended) is trailing operators and two indents on the the next line too separate the boolean expression from the body.

eric-wieser replied:
Is there something like CPEP 8 for cpython style guidelines?

MSeifert04 replied:
You mean [PEP7](https://www.python.org/dev/peps/pep-0007/)?

charris replied:
Also `doc/C_STYLE_GUIDE.rst.txt`. Here, second line just two indents from the `if`.

eric-wieser replied:
Already fixed here, I think

eric-wieser committed:
BUG: Allow array with 0 dims to be resized

eric-wieser commented:
Rebased to solve conflicts.

charris commented on numpy/core/src/multiarray/descriptor.c:
```diff
@@ -348,14 +350,21 @@ _convert_from_tuple(PyObject *obj)
         }
-        newdescr->elsize = type->elsize * items;
+        newdescr->elsize = nbytes;
         if (newdescr->elsize == -1) {
```
Can `nbytes` be -1? If so, could use a note. I'm guessing a single item of size 0, which seems restrictive.

eric-wieser replied:
Huh, that's pretty weird. This couldn't happen before either, could it?

charris commented on numpy/core/src/multiarray/getset.c:
```diff
@@ -469,22 +469,18 @@ array_descr_set(PyArrayObject *self, PyObject *arg)
-            Py_DECREF(newtype);
+    /*
+     * Treat V0 as resizable void - unless the destination is already V0, then
+     * don't allow np.void to be duplicated
```
Why not?

eric-wieser replied:
For the same reason that `np.dtype(np.void) is np.dtype(np.void)` is desirable. Without this special case, `v0_arr.view(np.void).dtype is not np.void`.

charris commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -1712,11 +1715,9 @@ array_setstate(PyArrayObject *self, PyObject *args)
-    }
-    if (size < 0 || size > NPY_MAX_INTP / PyArray_DESCR(self)->elsize) {
+    overflowed = npy_mul_with_overflow_intp(
+        &nbytes, size, PyArray_DESCR(self)->elsize);
```
Doesn't `size` need to be >= 0 in order to use `npy_mul_with_overflow_intp`? Seems it should be checked first to avoid a function call.

charris replied:
Could do something like
```
invalid = size < 0 || npy_mul_with_overflow_intp(...)
```

eric-wieser replied:
No need, there's already a explicit `a != 0` case in that function.

There's no point avoiding the function call, because optimizing the performance of empty arrays is pretty pointless

eric-wieser replied:
Oh I see, the issue is size == -1. Maybe I'll just duplicate the below `if`.

charris commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -1822,7 +1823,7 @@ array_setstate(PyArrayObject *self, PyObject *args)
             if (swap) {
                 /* byte-swap on pickle-read */
-                npy_intp numels = num / PyArray_DESCR(self)->elsize;
+                npy_intp numels = PyArray_SIZE(self);
```
Seems sensible, but why was it done differently before?

eric-wieser replied:
To avoid recalculating `num`, most likely

charris commented on numpy/core/src/multiarray/shape.c:
```diff
@@ -111,11 +112,11 @@ PyArray_Resize(PyArrayObject *self, PyArray_Dims *newshape, int refcheck,
 
-        if (newsize == 0) {
-            sd = PyArray_DESCR(self)->elsize;
+        if (newnbytes == 0) {
```
Could do
```
sd = newnbytes ? newnbytes : elsize
```

eric-wieser replied:
Would prefer to keep the `== 0`, because that's handling the special meaning of `0` which is "allow resizing".

Could flatten the `if` though

charris commented on numpy/core/src/multiarray/shape.c:
```diff
@@ -142,7 +142,7 @@ PyArray_Resize(PyArrayObject *self, PyArray_Dims *newshape, int refcheck,
         }
         else{
-            memset(PyArray_BYTES(self)+oldsize*elsize, 0, (newsize-oldsize)*elsize);
+            memset(PyArray_BYTES(self) + oldnbytes, 0, newnbytes - oldnbytes);
```
`newnbytes - oldnbytes` >= 0?

eric-wieser replied:
Yep, guaranteed by the enclosing `if`

charris commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -1201,6 +1201,74 @@ def test_count_nonzero_unaligned(self):
 
 
+class TestZeroSizeFlexible(object):
+    @staticmethod
```
Is there an advantage to making this a staticmethod?

eric-wieser replied:
To document that it doesn't use `self`?

charris replied:
Well, it does that, don't know if it is worth it as the normal impulse would be to use it only if needed. Then again, my Python experience started when Python 21 was the newest and greatest, so programming conventions may have changed. In fact, it is the latter possibility that interests me. Do you know if it is the case that this sort of programming style is being taught or encouraged?

eric-wieser replied:
`TestCase` objects (or the inferred use of them by pytest) are pretty weird anyway, because an new one is instantiated for each method call by the test runner. I can change this if you feel strongly.

charris replied:
I don't feel strongly about it. On occasion I find your coding style a bit unusual, but not wrong. Just trying to figure out if it is a personal style, or the future of python coding ;)

eric-wieser replied:
Could easily be personal style. I guess it probably comes from my java background, but even there it might just have been personal style.

charris commented:
Looks pretty good, just a couple of questions.

charris commented:
A release note under "Improvements" would be helpful. Someone might want to know just when the new functionality became available.

eric-wieser committed:
MAINT: Remove dead code

This could never be negative anyway, since elsize >= 0, and we already check for items >= 0.

eric-wieser committed:
BUG: Free dtype and dim objects correctly.

Previously `newdescr` was often leaked

eric-wieser committed:
DOC: Add release notes for 0-size dtypes

eric-wieser commented:
Caught a memory leak on the error path here too

charris commented:
Thanks Eric.
